engine/apps/mobile_app/tasks.py

```diff
@@ -153,78 +184,189 @@ def _get_fcm_message(alert_group, user, registration_id, critical):
+    if not device_to_notify:
+        _create_error_log_record()
+        logger.error(f"Error while sending a mobile push notification: user {user_pk} has no device set up")
+        return
+
+    message = _get_alert_group_escalation_fcm_message(alert_group, user, device_to_notify, critical)
+    _send_push_notification(device_to_notify, message, _create_error_log_record)
+
+
+def should_we_send_going_oncall_push_notification(
+    now: timezone.datetime, user_settings: "MobileAppUserSettings", schedule_event: ScheduleEvent
+) -> bool:
+    NOTIFICATION_TIMING_BUFFER = 15 * 60  # 15 minutes in seconds
+
+    # this _should_ always be positive since final_events is returning only events in the future
+    seconds_until_shift_starts = math.floor((schedule_event["start"] - now).total_seconds())
+
+    user_wants_to_receive_info_notifications = user_settings.info_notifications_enabled
+    # int representing num of seconds before the shift starts that the user wants to be notified
+    user_notification_timing_preference = user_settings.going_oncall_notification_timing
+
+    if not user_wants_to_receive_info_notifications:
+        logger.info("not sending going oncall push notification because info_notifications_enabled is false")
+        return False
+
+    lower_limit = seconds_until_shift_starts - NOTIFICATION_TIMING_BUFFER
+
+    timing_logging_msg = (
+        f"lower_limit: {lower_limit}\n"
+        f"user timing preference: {user_notification_timing_preference}\n"
+        f"seconds_until_shift_starts: {seconds_until_shift_starts}"
+    )
+
+    if lower_limit <= user_notification_timing_preference <= seconds_until_shift_starts:
+        logger.info(f"timing is right to send going oncall push notification\n{timing_logging_msg}")
+        return True
+    logger.info(f"timing is not right to send going oncall push notification\n{timing_logging_msg}")
+    return False
+
+
+@shared_dedicated_queue_retry_task(autoretry_for=(Exception,), retry_backoff=True, max_retries=MAX_RETRIES)
+def conditionally_send_going_oncall_push_notifications_for_schedule(schedule_pk) -> None:
+    # avoid circular import
+    from apps.mobile_app.models import MobileAppUserSettings
+
+    logger.info(f"Start calculate_going_oncall_push_notifications_for_schedule for schedule {schedule_pk}")
+
+    try:
+        schedule: OnCallSchedule = OnCallSchedule.objects.get(pk=schedule_pk)
+    except OnCallSchedule.DoesNotExist:
+        logger.info(f"Tried to notify user about going on-call for non-existing schedule {schedule_pk}")
+        return
+
+    now = timezone.now()
+
+    for schedule_event in schedule.final_events("UTC", now, days=5):
+        users = schedule_event["users"]
+
+        for user in users:
+            user_pk = user["pk"]
+            logger.info(f"Evaluating if we should send push notification for schedule {schedule_pk} for user {user_pk}")
+
+            try:
+                user = User.objects.get(public_primary_key=user_pk)
```

Comment:
You will be iterating over multiple events, and potentially the same user will appear several times. Would it be worth it to keep a dict/cache with already fetched users (and device? and maybe already notified check?) instead of going to the DB each time?

Reply:
good idea 👍  updated in [3335209](https://github.com/grafana/oncall/pull/1814/commits/3335209caeee989c25bb7c59a6c53e1bf5ed7684)